From 77e4f68dc48283100f3e9ebb26de407456bdb4d3 Mon Sep 17 00:00:00 2001
From: Thomas Colliers <mail@thomascolliers.com>
Date: Mon, 10 Aug 2026 22:37:17 +0200
Subject: [PATCH 3/5] media: qcom: camss: add milos (SM7635) resources and
 compatible

Describe the milos CAMSS complex: 4 CSIPHY, 3 CSID (reusing the gen2 RDI
ops), 3 TFE665, their clocks, interconnects and power domains, and add
the qcom,milos-camss compatible. The soc_ahb clock (CAM_CC_SOC_AHB_CLK)
gates the register-access AHB bus and is required for the ISP to be
reachable at all.

Signed-off-by: Thomas Colliers <mail@thomascolliers.com>
---
 drivers/media/platform/qcom/camss/camss.c | 123 ++++++++++++++++++++++
 1 file changed, 123 insertions(+)

diff --git a/drivers/media/platform/qcom/camss/camss.c b/drivers/media/platform/qcom/camss/camss.c
index 9335636d7..f21ca1659 100644
--- a/drivers/media/platform/qcom/camss/camss.c
+++ b/drivers/media/platform/qcom/camss/camss.c
@@ -820,6 +820,116 @@ static const struct resources_icc icc_res_2290[] = {
 	},
 };
 
+/*
+ * Milos (SM7635) resources. Camera front-end is TFE-based (TFE665/CSID665/
+ * CSIPHY v2.2.1). For the interfaces used here, CSID665 and CSIPHY reuse the
+ * QCM2290 csid_ops_340 and csiphy_ops_3ph_1_0 implementations. The TFE RDI
+ * path uses vfe_ops_665 for its different BUS offset and packer format.
+ * All clocks come from camcc-milos.
+ */
+#define MILOS_CSIPHY_RES(n)						\
+	{								\
+		.regulators = {},					\
+		.clock = { "cpas_ahb", "soc_ahb", "csiphy" #n,		\
+			   "csiphy" #n "_timer" },			\
+		.clock_rate = { { 0 },						\
+				{ 0 },						\
+				{ 240000000, 341330000, 384000000 },		\
+				{ 100000000, 200000000, 268800000 } },		\
+		.reg = { "csiphy" #n },						\
+		.interrupt = { "csiphy" #n },					\
+		.csiphy = {							\
+			.id = n,						\
+			.hw_ops = &csiphy_ops_3ph_1_0,				\
+			.formats = &csiphy_formats_sdm845			\
+		}								\
+	}
+
+static const struct camss_subdev_resources csiphy_res_milos[] = {
+	MILOS_CSIPHY_RES(0),
+	MILOS_CSIPHY_RES(1),
+	MILOS_CSIPHY_RES(2),
+	MILOS_CSIPHY_RES(3),
+};
+
+#define MILOS_CSID_RES(n)						\
+	{								\
+		.regulators = {},						\
+		.clock = { "cpas_ahb", "soc_ahb", "tfe" #n "_ahb", "csi" #n,	\
+			   "vfe" #n "_cphy_rx", "vfe" #n },			\
+		.clock_rate = { { 0 },						\
+				{ 0 },						\
+				{ 0 },						\
+				{ 192000000, 240000000, 384000000, 426400000 },	\
+				{ 0 },						\
+				{ 0 } },					\
+		.reg = { "csid" #n },						\
+		.interrupt = { "csid" #n },					\
+		.csid = {							\
+			.hw_ops = &csid_ops_340,				\
+			.parent_dev_ops = &vfe_parent_dev_ops,			\
+			.formats = &csid_formats_gen2				\
+		}								\
+	}
+
+static const struct camss_subdev_resources csid_res_milos[] = {
+	MILOS_CSID_RES(0),
+	MILOS_CSID_RES(1),
+	MILOS_CSID_RES(2),
+};
+
+#define MILOS_VFE_RES(n)						\
+	{								\
+		.regulators = {},						\
+		.clock = { "cpas_ahb", "soc_ahb", "tfe" #n "_ahb", "vfe" #n,	\
+			   "csi" #n, "vfe" #n "_cphy_rx",		\
+			   "camnoc_rt_axi", "camnoc_nrt_axi",		\
+			   "camnoc_axi_hf", "camnoc_axi_sf" },		\
+		.clock_rate = { { 0 },						\
+				{ 0 },						\
+				{ 0 },						\
+				{ 19200000, 153600000, 192000000, 256000000,	\
+				  384000000, 460800000, 600000000 },		\
+				{ 0 },						\
+				{ 0 },						\
+				{ 150000000, 240000000, 300000000, 400000000 },	\
+				{ 0 },						\
+				{ 0 },						\
+				{ 0 } },					\
+		.reg = { "vfe" #n },						\
+		.interrupt = { "vfe" #n },					\
+		.vfe = {							\
+			.line_num = 4,						\
+			.hw_ops = &vfe_ops_665,					\
+			.formats_rdi = &vfe_formats_rdi_845,			\
+			.formats_pix = &vfe_formats_pix_845			\
+		}								\
+	}
+
+static const struct camss_subdev_resources vfe_res_milos[] = {
+	MILOS_VFE_RES(0),
+	MILOS_VFE_RES(1),
+	MILOS_VFE_RES(2),
+};
+
+static const struct resources_icc icc_res_milos[] = {
+	{
+		.name = "ahb",
+		.icc_bw_tbl.avg = 150000,
+		.icc_bw_tbl.peak = 300000,
+	},
+	{
+		.name = "hf_mnoc",
+		.icc_bw_tbl.avg = 2097152,
+		.icc_bw_tbl.peak = 3000000,
+	},
+	{
+		.name = "sf_mnoc",
+		.icc_bw_tbl.avg = 2097152,
+		.icc_bw_tbl.peak = 3000000,
+	},
+};
+
 static const struct camss_subdev_resources csiphy_res_660[] = {
 	/* CSIPHY0 */
 	{
@@ -5148,6 +5258,18 @@ static const struct camss_resources msm8996_resources = {
 	.vfe_num = ARRAY_SIZE(vfe_res_8x96),
 };
 
+static const struct camss_resources milos_resources = {
+	.version = CAMSS_7635,
+	.csiphy_res = csiphy_res_milos,
+	.csid_res = csid_res_milos,
+	.vfe_res = vfe_res_milos,
+	.icc_res = icc_res_milos,
+	.icc_path_num = ARRAY_SIZE(icc_res_milos),
+	.csiphy_num = ARRAY_SIZE(csiphy_res_milos),
+	.csid_num = ARRAY_SIZE(csid_res_milos),
+	.vfe_num = ARRAY_SIZE(vfe_res_milos),
+};
+
 static const struct camss_resources qcm2290_resources = {
 	.version = CAMSS_2290,
 	.csiphy_res = csiphy_res_2290,
@@ -5321,6 +5443,7 @@ static const struct of_device_id camss_dt_match[] = {
 	{ .compatible = "qcom,msm8953-camss", .data = &msm8953_resources },
 	{ .compatible = "qcom,msm8996-camss", .data = &msm8996_resources },
 	{ .compatible = "qcom,qcm2290-camss", .data = &qcm2290_resources },
+	{ .compatible = "qcom,milos-camss", .data = &milos_resources },
 	{ .compatible = "qcom,qcs8300-camss", .data = &qcs8300_resources },
 	{ .compatible = "qcom,sa8775p-camss", .data = &sa8775p_resources },
 	{ .compatible = "qcom,sc7280-camss", .data = &sc7280_resources },
-- 
2.55.0

